### ISSUE TITLE
Proposal: AsyncClient API unification

### ISSUE THREAD

**Issue by pquentin** [encode/httpx#258]
To fully support HTTP/2, httpx will want to support multiplexing, arguably the most important HTTP/2 feature:

> For both SPDY and HTTP/2 the killer feature is arbitrary multiplexing on a single well congestion controlled channel. It amazes me how important this is and how well it works. One great metric around that which I enjoy is the fraction of connections created that carry just a single HTTP transaction (and thus make that transaction bear all the overhead). For HTTP/1 74% of our active connections carry just a single transaction - persistent connections just aren't as helpful as we all want. But in HTTP/2 that number plummets to 25%. That's a huge win for overhead reduction. Let's build the web around that. [(HTTP/2 is Live in Firefox, Patrick McManus)](http://bitsup.blogspot.com/2015/02/http2-is-live-in-firefox.html)

## Existing `parallel()` API

Since HTTP/2 is multiplexed, it means that clients such as httpx should only open [one connection per origin](https://http2.github.io/faq/#why-just-one-tcp-connection), even (especially?) for concurrent requests. This is indeed possible with the proposed `parallel()` API:

```python3
client = httpx.AsyncClient()
async with client.parallel() as parallel:
    pending_one = await parallel.get('https://example.com/1')
    pending_two = await parallel.get('https://example.com/2')
    response_one = await pending_one.get_response()
    response_two = await pending_two.get_response()
```

In that case, the TCP connection to example.com can live in an async task owned by the `parallel` object, receive orders with `await parallel.get()` and return responses with the `await pending.get_response()` calls. That works well, but I believe four things could be improved here:

 1. The APIs for performing serial requests and for performing parallel requests are different, so one must choose carefully, and evolving one's code from one API to the other requires work.
 2. And since most people don't read the docs, they're likely to launch requests in parallel using the client directly, and likely won't notice that their code does not take advantage of HTTP/2 multiplexing.
 3. Launching tasks is not the job of the HTTP client: that should be left to each async framework, since they all have a preferred style.
 4. The parallel API increases the API surface of httpx.

## Proposed unified async API

Based on those observations, I believe a better API would be to only allow instantiating the client using a context manager, eg. `async with httpx.AsyncClient() as client`. This then allows different styles. I'm not very familiar with asyncio, but I believe the above example would become:

```python3
async with httpx.AsyncClient() as client:
    response_one = asyncio.create_task(client.get('https://example.com/1'))
    response_two = asyncio.create_task(client.get('https://example.com/2'))
    await response_one
	await response_two
```

But you can also use other primitives, such as `asyncio.gather`:

```python3
async with httpx.AsyncClient() as client:
    for response in await asyncio.gather(
            client.get('https://example.com/1'),
            client.get('https://example.com/2')]:
        ...
```

And this fits more easily with other async frameworks, such as trio:

```python3
async with httpx.AsyncClient("trio") as client:
    async with trio.open_nursery() as nursery:
        nursery.start_soon(client.get, 'https://example.com/1')
        nursery.start_soon(client.get, 'https://example.com/2')
```

And it would also be the preferred way to launch single requests:

```
async with httpx.AsyncClient() as client:
    response = await client.get('https://example.com/1')
```

## What about the sync client?

The same logic applies for the sync client, and I would personally reuse what Python already offers to perform tasks in parallel: [concurrent.futures](https://docs.python.org/3/library/concurrent.futures.html). I guess an executor specific to httpx using asyncio behind the scenes could work.

Of course, it makes sense to keep `http.get(url)` for backwards compatibility with requests and to deal with the common case where users only need to make a single sync request.

## Conclusion

I believe this new proposed API has the following advantages:

 1. it's more unified (only one way to do it)
 2. it's natural to go from serial requests to concurrent requests
 3. it delegates parallel task creation to the async framework
 4. it reduces the API surface

The drawbacks are that the sync/async cases no longer use the same `parallel()` API and that the simple common async case is slightly more cumbersome to type, but I personally believe that this proposal is a better compromise.

What do you think?

(Disclaimer: this idea is originally from @njsmith, I took the time to turn it into a proposal and probably add errors of my own.)

**Comment by tomchristie** [comment]
> To fully support HTTP/2, httpx will want to support multiplexing, arguably the most important HTTP/2 feature.

We do.

If you're using multiple tasks, or multi threading, with a single client, then any HTTP/2 connections will already be multiplexed. In eg. a web app environment you don't really *see* that because the server is handling the concurrency aspect for you, so each *individual* code path reads as a single sequential request, but you'll actually end up having multiplexed requests across the same client. And yes, the AsyncClient can be used with any standard concurrency primitives (for whichever backend).

(Proviso: I think we may have some niggly races that aren't yet resolved in the threaded case, but that's an "we're in alpha" buglette, rather than an interface issue.)

Also "instantiating the client using a context manager, eg. async with httpx.AsyncClient() as client" isn't the right level here. Eg. supposing we're in a web app, then...

```
async def homepage():
    async with httpx.AsyncClient() as client:
        ... # Load a couple of resources in parallel
    ... # Return a response
```

Is the wrong thing to do, because you want to make sure you're using shared connection pooling across all the incoming requests, rather than just within the context of a single endpoint during a single request/response cycle. So what you actually want is...

```
async def homepage():
    async with TrioNurseryOrAsyncioSupervisorOrClientParallelContextOrWhatever:
        ... # Load a couple of resources in parallel using a globally shared `client` instance.
    ... # Return a response
```

What this issue *actually* reduces to is "let's not introduce the parallel requests API".

That's *feasible* although there's two primary reasons why we might want the parallel requests API...

* Providing async concurrency but from within a standard threaded interface. (This is the big one really, we can give users a really lightweight way of taking advantage of concurrent requests but within standard threaded codebases.)
* We *really* only want users to be using structured concurrency styles for branching. The parallel requests API meets that constraint, and stops folks using weaker asyncio primitives. (This is the weak of the two reasons. and could be addressed in other ways, in particular by async adopting an equivelent to "nursery", and starting to educate the ecosystem to prefer structured concurrency styles wherever possible)

I think there's also a *broader* issue here, around context manged vs. non context managed APIs. In particular:

* Client instances. (This would ensure that connection pools are strictly closed off, once a client is out of scope. Practically most enviroments will actually just want a single client for the lifetime of the application, so it's probably pretty useful to provide *either* context managed *or* unmanaged styles.)
* Streaming responses. (Right now we're following the requests API, which gives *either* context managed *or* unmanaged styles. The unmanaged style is a really gotcha there to my eyes, as it's super easy to create and not close streaming responses. That probably doesn't end up being a hard bug in practice, but rather just an unseen drag on resources or whatever)

I think that we should close the issue off, but not start or consider #52 until we've fully addressed the prerequisities of making sure that we've got support for both an asyncio and a trio backend, and that we're *all* happy whatever ConcurrencyBackend interfaces are neccessary in order to adequately support that.

**Comment by njsmith** [comment]
With the current API, if you're using a single client, it can't use HTTP/2 at all in a structured concurrency system like Trio, because HTTP/2 needs a background task and Trio won't let you make a background tasks without an `async with`.

**Comment by florimondmanca** [comment]
I'm not sure myself of how HTTP/2 multiplexing works/what it requires to have, but what is currently preventing users from using the API described by @pquentin, even if we have a single client?

```python
import functools
from .resources import client

async def do_stuff():
    async with trio.open_nursery() as nursery:
        nursery.start_soon(client.get, "https://example.com/1")
        nursery.start_soon(
            functools.partial(client.post, json={"foo": "bar"}),
            "https://example.com/2"
        )
    return PlainTextResponse("Done")
```

**Comment by tomchristie** [comment]
> because HTTP/2 needs a background task and Trio won't let you make a background tasks without an async with

HTTP/2 doesn't need a background task in order to support multiplexing, no. Any API interaction that gets a response, or streams some response content will end up progressing *any* active streams (although that's unseen from a users' perspective).

**Comment by njsmith** [comment]
@tomchristie HTTP/2 needs a background task whether you do multiplexing or not :-). Any open connection has to have some task listening for PING frames at all times, whether anyone's interacting with it or not.

**Comment by tomchristie** [comment]
Still not absolutely convinced that's neccessary. You'll handle a PING frame so long as some part of the code base, somewhere, is doing something with a request/response.

If nothing's interacting with it for a substantial enough period of time then then we'll never respond to the PING and the connection'll get dropped, which is reasonable behavior.

It *is* legit that we *could* want to have a background task associated with a client instance (both in order to handle PING frames, and to expire dropped connections or expired old keep-alive HTTP/1.1 connections) but it's not clear that it needs to be a hard requirement (or even *neccessarily* something we'd want to end up doing.)

**Comment by njsmith** [comment]
I dunno, I'm mostly relying on @lukasa here, but he spent some effort pounding it into my head that HTTP/2 definitely needed a task reading from the underlying connection at all times.

I don't think letting whichever task happens to be interacting with the connection do the driving is going to work in any case, because of cancellation. If your data send operation gets cancelled in the middle, it's extremely difficult to recover in any reasonable way. If it's in a background task where cancellation means that the whole `async with open_session` block is getting closed down, then that's OK, but if cancelling a `get` call can cause the underlying connection to get corrupted then that's no good.

We've spent a *lot* of time trying to figure out these patterns over the last few years, with simpler protocols like TLS and websocket. I originally thought like you are now, but I discovered I was wrong :-). Trio's TLS code does use the pattern you describe. That's a much simpler protocol flow-control wise – basically just two unidirectional streams that barely interact – and I think it's about at the limit of what you can handle that way; it took a ton of effort to get working and it still has some edge cases around cancellation that I'm not quite sure we're handling right.

**Comment by tomchristie** [comment]
Sure thing. I think we're doing the right thing at this point in time to keep progressing towards a mature HTTP/1.1 + HTTP/2 client. Eg. We've still got basics like #236 and #262 to work on.

If we *did* strictly want to move to a "clients must *always* be context-managed" style, then we *could*, but I'd rather have that driven by neccessity, even if that's only so that it helps fully thrash out a robust understanding of the issues there. (Eg. not really clear to me why background-task-for-dealing-with-pings helps any problems that task-cancellation-during-requests could cause) 

I think sensible things here *at this point in time* are:

* Defer #52 for the moment - we've got far more fundamental stuff to get sorted first in any case.
* Get a functional trio implementation in.

**Comment by njsmith** [comment]
> We've still got basics like #236 and #262 to work on.

Yeah, this is why we decided to defer HTTP/2 until after we had a viable HTTP/1.1 release.

> not really clear to me why background-task-for-dealing-with-pings helps any problems that task-cancellation-during-requests could cause

It's not the pings per se. The idea is to split things up so there are background tasks handling socket io, and the foreground tasks interact with them via some kind of queue. Socket IO in general is not atomic wrt cancellation: if a write gets interrupted half-way through, then the connection gets left in some indeterminate corrupted state. OTOH putting things in and out of queues is easy to make atomic. So this makes it so when a foreground task is cancelled, it can't leave things in an indeterminate state.

**Comment by tomchristie** [comment]
> Yeah, this is why we decided to defer HTTP/2 until after we had a viable HTTP/1.1 release.

Sure thing. We have the *option* of switching HTTP/2 to off-by-default for a 1.0 release if we want it, but we’re currently opting to keep it on, and deal with the bugs as they come in.

> So this makes it so when a foreground task is cancelled, it can't leave things in an indeterminate state.

That makes more sense, yup. I’d probably still be okay with treating backgrounding there as YAGNI. Eg we could just have a blunt policy of tearing down the entire connection if tasks get cancelled. I guess there might also be potentially sensible things we could do with shielding.

If we *do* find we’ve a hard requirement to switch to a background and queue model, then we can refactor to that in due course.

Seem reasonable?

**Comment by njsmith** [comment]
> That makes more sense, yup. I’d probably still be okay with treating backgrounding there as YAGNI. Eg we could just have a blunt policy of tearing down the entire connection if tasks get cancelled. I guess there might also be potentially sensible things we could do with shielding.

I don't think breaking unrelated tasks when one task is cancelled is going to work in the long run... and shielding isn't very useful here, because anytime you're dealing with the network, you have to let people impose timeouts, and those are a version of cancellation.

> If we do find we’ve a hard requirement to switch to a background and queue model, then we can refactor to that in due course.

I get what you're saying. I just want to point out that in this case, the refactor would involve breaking the public API for almost all users, so hopefully you're taking that into account in your planning :-)

**Comment by sethmlarson** [comment]
When you say public API do you mean the `ConcurrencyBackend` or more than that? To me the `ConcurrencyBackend` isn't really "user-facing" API. Sure it faces some users but not the ones that are just trying to make HTTP requests.

**Comment by pquentin** [comment]
@sethmlarson My understanding is that we're discussing whether httpx should mandate using `async with` to create a new `AsyncClient`, which is as user-facing as it can get. :)

My perspective is that there are many things that can get fixed in due time, but this specific bit of API cannot, and I would not want httpx to paint itself in a corner. Before structured concurrency, the solution would have been "oh, just start a task in the background, without an owner", but we now know that it can only lead to more trouble down the road.

The plan to experiment without this background task makes complete sense, but I personally hope that by 1.0 httpx will mandate `async with`, *especially* if the `parallel()` API is dropped.

@florimondmanca To answer [your question above](https://github.com/encode/httpx/issues/258#issuecomment-523387661), nothing prevents users to do this, but it does not help much either. It means that an exception inside that nursery block will cancel all the tasks it owns (which is good!), but other users of the client could still see corrupted state, and the client itself won't be able to take advantage of the nursery to start a background task (and get all of the advantages of using nurseries).

**Comment by sethmlarson** [comment]
You can see how much I understand about all this. :) This would definitely be something I'd like resolved before we tag a 1.X.
